mesonbuild/utils/vsenv.py

```diff
@@ -93,7 +94,8 @@ def _setup_vsenv(force: bool) -> bool:
     bat_file.write(bat_contents)
     bat_file.flush()
     bat_file.close()
-    bat_output = subprocess.check_output(bat_file.name, universal_newlines=True)
+    bat_output = subprocess.check_output(bat_file.name, universal_newlines=True,
+                                         encoding=locale.getpreferredencoding(False))
```

Comment:
Technically the default here is `encoding='locale'` (identical to `encoding=None` except with the warning squelched) but that was only added in python 3.10.

The difference is if you set [`PYTHONUTF8=1`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUTF8) before running meson then locale is still locale but getpreferredencoding is utf-8.

So maybe just check if `sys.version_info >= (3, 10)` and pass encoding='locale' as an unpacked dict of kwargs?

Reply:
According to the docs (https://docs.python.org/3/library/subprocess.html#frequently-used-arguments) and looking at the code (https://github.com/python/cpython/blob/aab6f7173a3b825599629dd6fa5cb7e477421595/Lib/subprocess.py#L375-L378) the default seems to be equivalent to `locale.getpreferredencoding(False)`.

I'm happy to adjust it to ignore utf-8 mode if wanted though.

Reply:
The default is *similar to*, not equal to.

There are a couple possibilities here.
- `encoding=None`
- `encoding='locale'`
- `encoding=getencoding()`
- `encoding=getpreferredencoding(False)`

The first three are all equivalent, because the first means to literally execute the third inside the internal implementation of subprocess. The second is a symbolic encoding that means to use the same encoding as whatever the third would produce. The third returns the name of a final encoding and doesn't include symbolism.

The fourth is adjacent to the third, and always returns the same value except for when utf-8 mode is enabled. The issue here is "except for", because if there are exceptions then it isn't equivalent, only "mostly equivalent".

As far as fixing it goes, the python ecosystem is in a bind:
- the first triggers an EncodingWarning
- the second and third require a minimum of python 3.10 or 3.11
- the fourth will change its behavior in python 3.15 and can opt in to changing its behavior starting with python 3.7


That being said, we already use the fourth in a couple places in meson, so it may just be fine to defer the issue until python 3.15 and then fix all of those issues at once.